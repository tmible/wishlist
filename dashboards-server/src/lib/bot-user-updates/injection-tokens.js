/** @module Токены внедрения адаптеров обновлений, полученных ботом */

/**
 * Токен внедрения хранилища
 * @constant {{}}
 */
export const Store = {};

/**
 * Токен внедрения сетевого сервиса
 * @constant {{}}
 */
export const NetworkService = {};

/**
 * Токен внедрения кэша
 * @constant {{}}
 */
export const Cache = {};
