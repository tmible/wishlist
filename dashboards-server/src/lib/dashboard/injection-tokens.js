/** @module Токены внедрения адаптеров дашборда */

/**
 * Токен внедрения фабричного метода хранилища
 * @constant {{}}
 */
export const StoreFactory = {};

/**
 * Токен внедрения фабричного метода функции запроса данных
 * @constant {{}}
 */
export const NetworkFactory = {};
