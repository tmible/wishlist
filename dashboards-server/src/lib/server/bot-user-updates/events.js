/** @module События обновлений, полученных ботом */

/**
 * Получение части обновлений
 * @constant {{}}
 */
export const GetBotUserUpdates = {};

/**
 * Подсчёт всех обновлений
 * @constant {{}}
 */
export const CountBotUserUpdates = {};
