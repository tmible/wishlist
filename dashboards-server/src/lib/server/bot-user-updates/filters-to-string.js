/**
 * @typedef {
 *   import('$lib/bot-user-updates/initialization.js').BotUserUpdatesFilters
 * } BotUserUpdatesFilters
 */

/**
 * Преобразование объекта фильтров к строке для использования в SQL выражении
 * @function filtersToString
 * @param {BotUserUpdatesFilters} filters Фильтры
 * @returns {string} SQL представление фильтров
 */
export const filtersToString = (filters) => {
  const filtersString = Object.entries(filters)
    .map(([ key, value ]) => {
      if (typeof value === 'object') {
        const filter = [];
        if (value.start) {
          filter.push(`${key} >= ${value.start}`);
        }
        if (value.end) {
          filter.push(`${key} <= ${value.end}`);
        }
        return filter.join(' AND ');
      }
      if (value === '') {
        return '';
      }
      return `${key} = ${value}`;
    })
    .filter(Boolean)
    .join(' AND ');
  return `${filtersString.length > 0 ? ' AND ' : ''}${filtersString}`;
};
