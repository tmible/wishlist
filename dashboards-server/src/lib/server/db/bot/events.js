/** @module События адаптера БД для бота */

/**
 * Получение времени ответа
 * @constant {{}}
 */
export const GetResponseTime = {};

/**
 * Получение времени обработки обновления
 * @constant {{}}
 */
export const GetProcessTime = {};

/**
 * Получение времени подотовки к обработке обновления
 * @constant {{}}
 */
export const GetStartupTime = {};

/**
 * Получение количества активных пользователей за день
 * @constant {{}}
 */
export const GetDAU = {};

/**
 * Получение количества активных пользователей за месяц
 * @constant {{}}
 */
export const GetMAU = {};

/**
 * Получение количества активных пользователей за год
 * @constant {{}}
 */
export const GetYAU = {};

/**
 * Получение доли успешно обработанных обновлений
 * @constant {{}}
 */
export const GetSuccessRate = {};
