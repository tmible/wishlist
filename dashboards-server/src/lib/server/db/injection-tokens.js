/** @module Токены внедрения адаптера БД */

/**
 * Токен внедрения БД
 * @constant {string}
 */
export const Database = 'database';
