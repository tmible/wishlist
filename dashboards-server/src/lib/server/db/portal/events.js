/** @module События адаптера БД для портала */

/**
 * Получение времени ответа
 * @constant {{}}
 */
export const GetResponseTime = {};

/**
 * Получение количества активных пользователей за день
 * @constant {{}}
 */
export const GetDAU = {};

/**
 * Получение количества активных пользователей за месяц
 * @constant {{}}
 */
export const GetMAU = {};

/**
 * Получение количества активных пользователей за год
 * @constant {{}}
 */
export const GetYAU = {};

/**
 * Получение доли успешно обработанных серверной частью запросов
 * @constant {{}}
 */
export const GetSuccessRate = {};

/**
 * Получение доли пользователей, прошедших аутентификацию после посещения лендинга
 * @constant {{}}
 */
export const GetAuthenticationFunnel = {};

/**
 * Получение количества полученных серверной частью за секунду запросов
 * @constant {{}}
 */
export const GetRPS = {};
