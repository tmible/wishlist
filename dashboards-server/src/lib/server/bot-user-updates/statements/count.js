import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { Database } from '$lib/server/db/injection-tokens.js';
import { CountBotUserUpdates } from '../events.js';
import { filtersToString } from '../filters-to-string.js';

/**
 * @typedef {
 *   import('$lib/bot-user-updates/initialization.js').BotUserUpdatesFilters
 * } BotUserUpdatesFilters
 */

/**
 * Создание и выполенение SQL выражения для подсчёта в БД с логами части полученных ботом обновлений
 * @function countBotUserUpdates
 * @param {number} timeLock Таймштамп, позже которого выбираются записи
 * @param {BotUserUpdatesFilters} filters Фильтры обновлений
 * @returns {number} Количество обновлений, удовлетворяющих всем параметрам
 */
const countBotUserUpdates = (timeLock, filters = {}) => inject(Database).prepare(`
  SELECT COUNT(*) AS total
  FROM 'bot.logs'
  WHERE level = 30 AND msg = 'start processing update' AND time < ?${filtersToString(filters)}
`).get(timeLock).total;

/**
 * Подписка [создания и выполнения SQL выражения]{@link countBotUserUpdates}
 * на соответствующее событие
 * @function initCountBotUserUpdatesStatement
 * @returns {void}
 */
export const initCountBotUserUpdatesStatement = () => {
  subscribe(CountBotUserUpdates, countBotUserUpdates);
};
