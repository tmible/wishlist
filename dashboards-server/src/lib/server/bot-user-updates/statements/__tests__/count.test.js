import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { Database } from '$lib/server/db/injection-tokens.js';
import { CountBotUserUpdates } from '../../events.js';
import { filtersToString } from '../../filters-to-string.js';
import { initCountBotUserUpdatesStatement } from '../count.js';

vi.mock('@tmible/wishlist-common/dependency-injector');
vi.mock('@tmible/wishlist-common/event-bus');
vi.mock('$lib/server/db/injection-tokens.js', () => ({ Database: 'database' }));
vi.mock('../../events.js', () => ({ CountBotUserUpdates: 'count bot user updates' }));
vi.mock('../../filters-to-string.js');

describe('bot user updates / statements / count', () => {
  let db;
  let statement;

  beforeEach(() => {
    statement = { get: vi.fn().mockReturnValue({ total: 'total' }) };
    db = { prepare: vi.fn(() => statement) };
    vi.mocked(inject).mockReturnValue(db);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should subscribe to event', () => {
    initCountBotUserUpdatesStatement();
    expect(
      vi.mocked(subscribe),
    ).toHaveBeenCalledWith(
      vi.mocked(CountBotUserUpdates),
      expect.any(Function),
    );
  });

  describe('subscriber', () => {
    let eventHandler;

    beforeEach(() => {
      vi.mocked(subscribe).mockImplementationOnce((event, handler) => eventHandler = handler);
      initCountBotUserUpdatesStatement();
    });

    it('should inject database', () => {
      eventHandler('timeLock', 'filters');
      expect(vi.mocked(inject)).toHaveBeenCalledWith(vi.mocked(Database));
    });

    it('should prepare statement', () => {
      vi.mocked(filtersToString).mockReturnValueOnce('filters string');
      eventHandler('timeLock', 'filters');
      expect(db.prepare.mock.calls[0]).toMatchSnapshot();
    });

    it('should run statement on event emit', () => {
      eventHandler('timeLock', 'filters');
      expect(statement.get).toHaveBeenCalledWith('timeLock');
    });

    it('should return statement run result on event emit', () => {
      expect(eventHandler('timeLock', 'filters')).toBe('total');
    });
  });
});
