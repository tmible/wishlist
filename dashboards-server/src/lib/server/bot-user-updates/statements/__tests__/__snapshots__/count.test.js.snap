// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bot user updates / statements / count > subscriber > should prepare statement 1`] = `
[
  "
  SELECT COUNT(*) AS total
  FROM 'bot.logs'
  WHERE level = 30 AND msg = 'start processing update' AND time < ?filters string
",
]
`;
